### PR TITLE
feat(internal): generate dynamic datasource imports

### PR DESCRIPTION
**Pull request by viceice**
Generate imports on type-check to fully type check datasource api.
Generated files are only for testing now.

blocked by: #5548

### CODE REVIEW

**Comment by ViceIce** [comment]
feedback welcome

**Comment by rarkins** [comment]
Can you explain it? :)

File: package.json [hunk 1]
```diff
@@ -7,12 +7,13 @@
...
+    "build": "run-s clean generate:* compile:* copy-static-files",
+    "build:docker": "run-s generate:* compile:ts copy-static-files",
     "clean": "rimraf dist",
     "clean-cache": "node bin/clean-cache.js",
     "compile:ts": "tsc -p tsconfig.app.json",
     "compile:dts": "tsc -p tsconfig.dts.json",
+    "generate:imports": "node --experimental-modules tools/generate-imports.js",
     "copy-static-files": "copyfiles -u 1 -e **/__fixtures__/**  -e **/__mocks__/** lib/**/*.json lib/**/*.py dist/",
     "create-json-schema": "babel-node --extensions \".ts,.js\" -- bin/create-json-schema.js && prettier --write \"renovate-schema.json\"",
     "debug": "babel-node --inspect-brk  --extensions \".ts,.js\" -- lib/renovate.ts",
```
Comment (zharinov) [inline package.json]: `"test": "run-s lint test-schema generate:imports type-check jest"`

**Comment by rarkins** [comment]
I had expected our best long term solution was making datasource into a class and then verifying it that way. Would this method be replaced by that if so?

**Comment by ViceIce** [comment]
Yes, this will partially removed, when we have classes. But the generated files allowes the typescript compiler and eslint to do static type checks.

File: package.json (hunk 1 above)
Comment (ViceIce) [reply package.json]: I'm on fixing build, thinking doing the build before lint.

**Comment by zharinov** [comment]
While considered as code smell, looks like code generation is preferable over runtime checks for such an important interface?

**Comment by ViceIce** [comment]
Another aproach would be to use the generated files only for tests and type check and exclude it for production builds.

**Comment by viceice** [comment]
New approach.

Generate imports but only for type-check. Would extend it if we want to go that way.
Maybe i can add a jest test too, to be sure generation works as expected.

**Comment by rarkins** [comment]
So this now just does the dynamic generation for the purposes of *checking* and isn't part of the regular build workflow?

**Comment by viceice** [comment]
yes

**Comment by viceice** [comment]
As we need compiletime references to all files for #5759 i've added the needed generation for datasources.

So @rarkins @JamieMagee  @zharinov I would like do discuss this solution. If all is right, i would do this for all other dynamic imports.

**Comment by rarkins** [comment]
How would this work during development? e.g. if I add a new datasource would I need to run `yarn build` before `yarn start ...` works?

**Comment by viceice** [comment]
I can add the generate step before start.

**Comment by viceice** [comment]
I can also improve to not write or delete files if content would not change.

**Pull request merged**